README
This is project OdbParser, created on Thu Sep  7 17:08:32 CEST 2006.

To compile and install OdbParser, type the following (as root):

python setup.py build
python setup.py install

The source files of OdbParser are distributed under the GNU Public License.


ChangeLog
2014-06-16  Morten Kjeldgaard  <mok@homunculus.local>

	* Fix compilation errors, add declarations in odb_io.h.
	* src/odbparsermodule.c: Ported to Python 3 API,


src/odb_io.c
/*
   Routines to read from binary O files.
   Morten Kjeldgaard, 03-Jan-2001.
   Copyright (C) Morten Kjeldgaard 2001-2006, 2014.
   Licence: GPL.
*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <ctype.h>
#include <inttypes.h> // define int32_t
#include "odb_io.h"

/*
  Swap bytes in n 4-byte words
*/
static void
swap4 (char *buffer, int n)
{
  register int i;
  char j;

  for (i=0; i < n*4; i+=4) {
    j = buffer[i];
    buffer[i] = buffer[i+3];
    buffer[i+3] = j;
    j = buffer[i+1];
    buffer[i+1] = buffer[i+2];
    buffer[i+2] =j;
  }
}

/*
  Read the parameter (datablock) from a binary O file.
*/
int read_param (int fd, char *par, char *partyp, int *size, int swap)
{
  int n;
  int32_t  rl1, rl2;

  n = read (fd, &rl1, 4);
  if (n == 0) return -1;
  if (swap) swap4 ((char *)&rl1, 1);
  n = read (fd, par, 25);
  // convert datablock name to lower case
  for (n=0; n<25; n++)
    par[n] = tolower(par[n]);
  n = read (fd, partyp, 1);
  n = read (fd, size, 4);
  if (swap) swap4 ((char *)size, 1);
  n = read (fd, &rl2, 4);
  if (swap) swap4 ((char *)&rl2, 1);

  if (rl1 != rl2) {
    fprintf (stderr, "Error reading parameter header (%d %d %d)\n",
	     *size, rl1, rl2);
    return -2;
  }
  return 0;
}

/*
  Read a text datablock from the binary file
*/
int read_text (int fd, char *text, int size, int swap)
{
  int n;
  int32_t rl1, rl2;

  n = read (fd, &rl1, 4);
  if (n == 0) return -1;
  if (swap) swap4 ((char *)&rl1, 1);
  n = read (fd, text, rl1);
  n = read (fd, &rl2, 4);
  if (swap) swap4 ((char *)&rl2, 1);

  if (rl1 != rl2) {
    fprintf (stderr, "Error read text block\n");
    return -2;
  }
  if (size != rl2)
    fprintf (stderr, "read_text: Expected %d, got %d elements\n", size, rl2);

  return 0;
}

/*
   Read 'size' C6 variables from the binary fortran file.
*/
int read_c6 (int fd, char *cstore, int size, int swap)
{
  int n;
  int32_t rl1, rl2;

  n = read (fd, &rl1, 4);
  if (n == 0) return -1;
  if (swap) swap4 ((char *)&rl1, 1);
  n = read (fd, cstore, rl1);
  n = read (fd, &rl2, 4);
  if (swap) swap4 ((char *)&rl2, 1);

  if (rl1 != rl2) {
    fprintf (stderr, "Error read character block\n");
    return -2;
  }
  if (6*size != rl2)
    fprintf (stderr, "read_c6: Expected %d, got %d elements\n", 6*size, rl2);

  return 0;
}

/*
  Read 'size' integers from the binary fortran file.  Swap bytes if
  necessary, file is always in big-endian order.
*/
int read_int4 (int fd, int *istore, int size, int swap)
{
  int n;
  int32_t rl1, rl2;

  n = read (fd, &rl1, 4);
  if (n == 0) return -1;
  if (swap) swap4 ((char *)&rl1, 1);
  n = read (fd, istore, rl1);
  if (swap) swap4 ((char *)istore, size);
  n = read (fd, &rl2, 4);
  if (swap) swap4 ((char *)&rl2, 1);

  if (rl1 != rl2) {
    fprintf (stderr, "Error read int block\n");
    return -2;
  }
  if (4*size != rl2)
    fprintf (stderr, "read_int4: Expected %d, got %d elements\n", 4*size, rl2);

  return 0;
}

/*
   Read 'size' floats from the binary fortran file.  Swap bytes if
   necessary, file is always in big-endian order.
*/
int read_float4 (int fd, float *rstore, int size, int swap)
{
  int n;
  int32_t rl1, rl2;

  n = read (fd, &rl1, 4);
  if (n == 0) return -1;
  if (swap) swap4 ((char *)&rl1, 1);
  n = read (fd, rstore, rl1);
  if (swap) swap4 ((char *)rstore, size);
  n = read (fd, &rl2, 4);
  if (swap) swap4 ((char *)&rl2, 1);

  if (rl1 != rl2) {
    fprintf (stderr, "Error read float block\n");
    return -2;
  }
  if (4*size != rl2)
    fprintf (stderr, "read_float4: Expected %d, got %d elements\n", 4*size, rl2);

  return 0;
}

/*
  Local Variables:
  mode: c
  mode: font-lock
  End:
*/


src/odb_io.h
/* 
   Copyright (C) Morten Kjeldgaard 2001-2006, 2014.
   License: GPL
*/

#if defined(MIPSEL) || defined(__i386__) || defined(__x86_64__) || defined(WIN32)
#  define DOSWAP 1
#else
#  define DOSWAP 0
#endif

/* Declaration of binary read functions */
int read_param (int fd, char *par, char *partyp, int *size, int swap);
int read_text (int fd, char *text, int size, int swap);
int read_c6 (int fd, char *cstore, int size, int swap);
int read_int4 (int fd, int *istore, int size, int swap);
int read_float4 (int fd, float *rstore, int size, int swap);

/* Declaration of formatted read functions */
int read_param_f (FILE *fp, char *par, char *partyp, int *size, char *fmt);
int read_int4_f (FILE *fp, int *array, int size);
int read_float4_f (FILE *fp, float *array, int size);
int read_c6_f (FILE *fp, char *array, int size, char *fmt);
int read_text_f (FILE *fp, char *array, int nrec, int size);

/*
  Local Variables: 
  mode: c
  mode: font-lock
  End:
*/


src/odb_io_f.c
/*
   Routines to read from formatted O files.
   Morten Kjeldgaard, 03-Jan-2001.
   Copyright (C) Morten Kjeldgaard 2001-2006, 2014.
   Licence: GPL.
*/

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>

#define MAXWORD 100

/* Get a word from a stream.  Maximum word size is MAXWORD
 * characters. If a word reaches it's maximum limit, we choose not to
 * flush the rest of the word. Returns NULL on EOF.  */
static char * getword(FILE *fp)
{
  static char word[MAXWORD + 1];
  char *p = word;
  int c;

  while ((c = fgetc(fp)) != EOF && isspace(c))
    ;                           /* Skip over word separators */
  if (c == EOF)
    return(NULL);
  *p++ = c;
  while ((c = fgetc(fp)) != EOF && !isspace(c) && p != &(word[MAXWORD])) {
    *p++ = c;                   /* Quit when a word separator is encountered
                                 * or we reach maximum word length
                                 */
  }
  *p = '\0';                    /* Mustn't forget that word terminator */
  return ((c == EOF) ? NULL : word);
}



/*
  Read the header of a formatted O file. Reads past any comments and
  blank lines
*/
int read_param_f (FILE *fp, char *par, char *partyp, int *size, char *fmt)
{
  char buf[256], *ch, *stat;

    // Get the first non-comment line.
  while (1) {
    if (!fgets(buf, 256, fp)) {
      return -1;
    }
    ch = strtok(buf," ");
    if (ch && ch[0] == '!')
      continue;
    if (ch && ch[0] == '\n')
      continue;
    break;
  }

  // Decode datablock name and convert to lower case
  memcpy (par, ch, 26);
  ch = par;
  while (*ch) {
    *ch = tolower(*ch);
    ch++;
  }

  // decode datablock type (single character == I, R, C, T)
  ch = strtok(NULL, " ");
  if (!ch)
    return 1;
  *partyp = *ch;

  // decode datablock size (in elements)
  ch = strtok(NULL, " ");
  if (!ch)
    return 2;
  *size = (int)strtol(ch, &stat, 10);
  if (*stat) {
    fprintf (stderr, "non-digits in datablock size\n");
    return 3;
  }

  // Finally, decode the format.
  ch = strtok(NULL, " \012");
  if (!fmt)
    return 4;
  memcpy (fmt, ch, 64);

  return 0;
}

/*
  Read 'size' integers from the file
*/

int read_int4_f (FILE *fp, int *array, int size)
{
  register int i;
  char *ch, *stat;

  for (i=0; i<size; i++) {
    ch = getword(fp);
    array[i] = (int)strtol(ch, &stat, 10);
    if (*stat) {
      fprintf (stderr, "non-digits in datablock\n");

      return 1;
    }
  }
  return 0;
}

/*
  Read 'size' floats from the file
*/
int read_float4_f (FILE *fp, float *array, int size)
{
  register int i;
  char *ch, *stat;

  for (i=0; i<size; i++) {
    ch = getword(fp);
    array[i] = (float)strtod(ch, &stat);
    if (*stat) {
      fprintf (stderr, "non-digits in datablock\n");
      return 1;
    }
  }
  return 0;
}

/*
  A simple parser for the fortran formats that the type C datablocks
  are stored in. Generate a string describing the format, so the
  program knows what to expect in the file. The function knows that
  type C variables are 6 bytes always. Examples:

  (2a)         -> 123456123456
  (1x,5a)      -> _123456123456123456123456123456
  (5(1x,a6))   -> _123456_123456_123456_123456_123456
  (1x,2(2x,a)) -> ___123456__123456
  (aaaaa)      -> 123456123456123456123456123456

*/
#define RSIZ 256

static char *parse_format(char *fmt)
{
  register int i;
  char *result, *f, *r, *t;
  int mult = 0;

  //printf ("allocating %d bytes\n",RSIZ);
  result = malloc(RSIZ);
  memset (result, 0, RSIZ);

  f = fmt;
  r = result;

  while (*f) {
    switch (*f) {
    case '(':
      t = parse_format(++f);
      if (mult == 0) mult = 1;
      for (i=0; i < mult; i++)
	strncat(result, t, RSIZ);
      free(t);
      //printf ("freed %d bytes\n",RSIZ);
      //printf ("size of result: %d\n",strlen(result));
      return result;
      break;
    case 'x':
    case 'X':
      if (mult == 0) mult = 1;
      for (i=0; i < mult; i++)
	*r++ = '_';
      mult = 0;
      break;
    case 'a':
    case 'A':
      if (mult == 0) mult = 1;
      for (i=0; i < mult; i++) {
	*r++ = '0'+1;
	*r++ = '0'+2;
	*r++ = '0'+3;
	*r++ = '0'+4;
	*r++ = '0'+5;
	*r++ = '0'+6;
      }
      mult = 0;
      break;
    case '0':
    case '1':
    case '2':
    case '3':
    case '4':
    case '5':
    case '6':
    case '7':
    case '8':
    case '9':
      mult = mult*10+(*f)-'0';
      //printf ("mult = %d\n", mult);
      break;
    case ')':
      //printf ("result:%s\n", result);
      return result;
    case ',':
      mult = 0;
      break;
    }
    f++;
  }
  fprintf (stderr, "shouldn't happen\n");
  return NULL;
}

/*
  Read 'size' C6 variables from the file
*/
int read_c6_f (FILE *fp, char *array, int size, char *fmt)
{
  register int i;
  char c, *a, *t, *s;
  int eol, inword;

  t = parse_format(fmt);
  s = t;
  a = array;

  i = 0;
  eol = 0;
  while (i < size) {
    //fprintf (stderr, "%d ", i);

    if (!*s) { // if end of format string, rewind it.
      s = t;   // and reset everything
      eol = 0;
      while (fgetc(fp) != '\n')
	;
      //fprintf (stderr, "rewind\n");
    }

    inword = 1;
    while (inword) {
      if (!eol)
	c = fgetc(fp);
      if (c == '\n')
	eol = 1;
      switch (*s++) {
      case '_':
	break;
      case '1':
      case '2':
      case '3':
      case '4':
      case '5':
	if (eol)
	  //*a++ = ' ';
	  *a++ = '#';
	else
	  *a++ = c;
	break;
      case '6':
	if (eol)
	  //*a++ = ' ';
	  *a++ = '6';
	else
	  *a++ = c;
	i++;
	inword = 0;
	break;
      }
    }  // end get one word
  }
  free(t);
  return 0;
}

/*
  Read a text datablock from the formatted file
*/
int read_text_f (FILE *fp, char *array, int nrec, int size)
{
  char buf[256];
  register int i;
  int j, n;

  for (i=0, j=0; i<nrec; i++) {
    fgets (buf, 256, fp);
    n = strlen(buf);
    strncpy (array+j, buf, size);
    j += size;
  }
  return j;
}

/*
  Local Variables:
  mode: c
  mode: font-lock
  End:
*/



src/odbparsermodule.c
/*
   Python module to read O data files, binary or formatted.
   Morten Kjeldgaard, 03-Jan-2001.
   Copyright (C) Morten Kjeldgaard 2001-2006, 2014.
   Licence: GPL.
*/

#include <Python.h>             /* Python header files */
#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
#include <arrayobject.h>
#include <fcntl.h>
#include "odb_io.h"

/*
  binfil -- return 1 if fnam is a binary O file, else 0. An O binary
  file normally has the byte pattern [0 0 0 036 .] in the first 5
  bytes of the file. */

static int binfil(char *fnam)
{
  int fd, n;
  char buf[8];

  fd = open(fnam, O_RDONLY);
  if (fd > -1) {
    n = read (fd, buf, 8);
    close(fd);
    if ((buf[0]&buf[1]&buf[2]) == 0 && buf[3] == 30 && buf[4] == '.')
      return 2;
    if ((buf[0]&buf[1]&buf[2]) == 0 && buf[3] == 30)
      /* The only binary O files that do not have a '.' in the fifth byte
	 are the dgnl data files. */
      return 1;
  }
   return 0;
}

/*
  Read a binary O database. The data is returned in a Python
  dictionary, with datablock names as keys. Real and integer data are
  stored in numpy arrays.  Type 'C' datablocks are in O character
  strings of length 6. These are returned as a tuple of strings. Type
  'T' datavblocks are returned as a tuple of strings.  Trailing spaces
  are stripped from both type 'C' and 'T' datablocks.
 */
static PyObject *readbinary (char *fnam)
{
  int fd;
  char par[26], typ, *s;
  int errcod, siz;
  npy_intp dims[] = {0};
  void *vector, *data;
  PyObject *pydict, *pykey, *pytup, *pystr;

  fd = open(fnam, O_RDONLY);
  if (fd < 0) {
    return NULL;
  }

  memset (par, 0, 26);
  pydict = PyDict_New();

  while (1) {

    errcod = read_param(fd, par, &typ, &siz, DOSWAP);
    if (errcod < 0 || siz == 0)
      break;
    //printf ("errcod=%d, param=%s type=%c size=%d\n", errcod, par, typ, siz);

    /* strip spaces off end of datablock name */
    s = &par[25];
    while (*s <= 32 && s > par)
      *s-- = '\0';

    pykey = PyUnicode_FromString(par);

    switch(typ) {

    case 'I':
      data = calloc(siz, sizeof(int));
      read_int4 (fd, data, siz, DOSWAP);
      dims[0] = siz;
      vector = (void *)PyArray_SimpleNewFromData(1, dims, NPY_INT, data);
      if (!vector) {
	close(fd);
	PyErr_SetString(PyExc_RuntimeError, "Failed to create integer array");
	Py_DECREF(pykey);	// decref latest key
	return pydict;		// return what we've got so far
      }
      PyDict_SetItem (pydict, pykey, (PyObject *)vector); // add to dictionary
      break;

    case 'R':
      data = calloc(siz, sizeof(float));
      read_float4 (fd, data, siz, DOSWAP);
      dims[0] = siz;
      vector = (void *)PyArray_SimpleNewFromData(1, dims, NPY_FLOAT, data);
      if (!vector) {
	close(fd);
	PyErr_SetString(PyExc_RuntimeError, "Failed to create real array");
	Py_DECREF(pykey);	// decref latest key
	return pydict;	// return what we've got so far
      }
      PyDict_SetItem (pydict, pykey, (PyObject *)vector); // add to dictionary
      break;

    case 'C':
      {
	register int i;
	char buf[7], *ch;
	buf[6] = 0;
	s = calloc (siz,6*sizeof(char));
	read_c6 (fd, s, siz, DOSWAP);
	pytup = PyTuple_New (siz);
	for (i=0; i < siz; i++) {
	  memcpy(buf, &s[6*i], 6);

	  /* strip spaces off end */
	  ch = &buf[6];
	  while (*ch <= 32 && ch > buf)
	    *ch-- = '\0';
	  /*
	     In principle we should use:
	     pystr = PyBytes_FromString(buf);
	     but it appears there are non-ascii bytes in some of
	     the character blocks in the distributed O data files.
	  */
	  pystr = PyBytes_FromString(buf);
	  if (PyTuple_SetItem (pytup, i, pystr) != 0)
	    fprintf (stderr, "tuple insert error");
	}
	free(s);
	PyDict_SetItem (pydict, pykey, pytup); // add to dictionary
      }
      break;

    case 'T':
      {
	register int i,j;
	char *ch, *t;
	int nrec=0;

	s = calloc(siz,sizeof(char));
	t = calloc(siz,sizeof(char)); // get a string that's big enough
	read_text (fd, s, siz, DOSWAP);

	for (i=0; i<siz; i++) { // count the number of records
	  if (s[i] == '\r')
	    nrec++;
	}
	pytup = PyTuple_New (nrec);

	nrec = 0;
	for (i=0,j=0; i<siz; i++) { // extract the individual strings into 't'
	  t[j] = s[i];

	  if (t[j] == '\r') {
	    ch = &t[j];
	    while (*ch <= 32 && ch > t) // strip spaces off end
	      *ch-- = '\0';

	    pystr = PyUnicode_FromString(t); // create python string
	    //pystr = PyBytes_FromString(t); // create python string
	    if (PyTuple_SetItem (pytup, nrec++, pystr) != 0) // add it to the tuple
	      fprintf (stderr, "tuple insert error");
	    j=0;
	  } else {
	    j++;
	  }
	}
	free(t);
	free(s);
	PyDict_SetItem (pydict, pykey, pytup); // add tuple to dictionary
      }

      break;

    } // end switch (typ)
  }
  close(fd);
  return pydict;
}

/*
   Read a formatted O datablock file. The current algorithm for
   reading formatted type 'C' datablocks requires that there are
   characters other than spaces in every element. This is in fact not
   always the case, for example the .major_menu datablock in menu.o
   has empty elements, which are filled with spaces by the Fortran
   FORMAT statement. It would require a lot of programming to deal
   with this issue, and it is frankly not important enough.
*/
static PyObject *readformatted (char *fnam)
{
  FILE *fp;
  char par[26], typ, fmt[64];
  int errcod, siz;
  npy_intp dims[] = {0};
  void *vector, *data;
  PyObject *pydict, *pykey, *pytup, *pystr;

  fp = fopen(fnam, "r");
  if (!fp) {
    return NULL;
  }

  pydict = PyDict_New();

  errcod = read_param_f(fp, par, &typ, &siz, fmt);
  while (!errcod) {

    pykey = PyUnicode_FromString(par);

    switch (toupper(typ)) {

    case 'I':
      data = calloc(siz, sizeof(int));
      read_int4_f (fp, data, siz);
      dims[0] = siz;
      vector = (void *)PyArray_SimpleNewFromData(1, dims, NPY_INT, data);
      if (!vector) {
	fclose(fp);
	PyErr_SetString(PyExc_RuntimeError, "Failed to create integer array");
	Py_DECREF(pykey);	// decref latest key
	return pydict;		// return what we've got so far
      }
      PyDict_SetItem (pydict, pykey, (PyObject *)vector); // add to dictionary

      break;

    case 'R':
      data = calloc(siz, sizeof(float));
      read_float4_f (fp, data, siz);
      dims[0] = siz;
      vector = (void *)PyArray_SimpleNewFromData(1, dims, NPY_FLOAT, data);
      if (!vector) {
	fclose(fp);
	PyErr_SetString(PyExc_RuntimeError, "Failed to create real array");
	Py_DECREF(pykey);	// decref latest key
	return pydict;	// return what we've got so far
      }
      PyDict_SetItem (pydict, pykey, (PyObject *)vector); // add to dictionary

      break;

    case 'C':
      {
	register int i;
	char buf[7], *ch, *s;
	buf[6] = 0;
	s = calloc (siz,6*sizeof(char));
	read_c6_f (fp, s, siz, fmt);
	pytup = PyTuple_New (siz);
	for (i=0; i < siz; i++) {
	  memcpy(buf, &s[6*i], 6);

	  /* strip spaces off end */
	  ch = &buf[6];
	  while (*ch <= 32 && ch > buf)
	    *ch-- = '\0';

	  pystr = PyUnicode_FromString(buf);
	  if (PyTuple_SetItem (pytup, i, pystr) != 0)
	    fprintf (stderr, "tuple insert error");
	}
	free(s);
	PyDict_SetItem (pydict, pykey, pytup); // add to dictionary
      }
      break;

    case 'T':
      {
	register int i,j;
	char *ch, *s, *t;
	int nrec, reclen;

	reclen = strtol(fmt, NULL, 10);
	nrec = siz;
	s = calloc(nrec*reclen,sizeof(char));
	t = calloc(reclen,sizeof(char)); // get a string that's big enough
	read_text_f (fp, s, nrec, reclen);

	pytup = PyTuple_New (nrec);

	for (i=0, j=0; i<nrec; i++) { // extract the individual strings into 't'
	  memcpy (t, s+j, reclen);
	  ch = &t[reclen-1];
	  while (*ch <= 32 && ch > t) // strip spaces off end
	    *ch-- = '\0';
	  pystr = PyUnicode_FromString(t); // create python string
	  if (PyTuple_SetItem (pytup, i, pystr) != 0) // add it to the tuple
	    fprintf (stderr, "tuple insert error");
	  j += reclen;
	}
	free(t);
	free(s);
	PyDict_SetItem (pydict, pykey, pytup); // add tuple to dictionary
      }

      break;
    } // end switch
    errcod = read_param_f(fp, par, &typ, &siz, fmt);
  }
  fclose(fp);
  return pydict;
}


/* 1. Functions available in odbparser module */

static PyObject *get (PyObject *self, PyObject *args)
{
  char *fnam;
  PyObject *pydict;

  if (!PyArg_ParseTuple(args, "s" , &fnam ))
    return NULL;

  /* Do the actual reading. The two subroutines readbinary and readformatted
     both return a Python dictionary. */

  if (binfil(fnam)) {
    //fprintf (stderr, "Reading binary O file\n");
    pydict = readbinary(fnam);
  }  else {
    //fprintf (stderr, "Reading formatted O file\n");
    pydict = readformatted(fnam);
  }
  return pydict;
}


/* 2. Doc strings */

static char odbparser_module__doc__[] =
"Parse O binary and formatted files";

static char odbparser_get__doc__[] =
"get(filename) -- return dictionary of O datablocks";


/* 3. Method table mapping names to wrappers */

static PyMethodDef odbparser_methods[] = {
  {"get", (PyCFunction)get,   METH_VARARGS, odbparser_get__doc__ },
  {NULL, (PyCFunction)NULL, 0, NULL} /* sentinel */
};


/* 4. Module initialization function */

static struct PyModuleDef moduledef = {
  PyModuleDef_HEAD_INIT,
  "odbparser",             /* name of module */
  odbparser_module__doc__, /* module documentation */
  -1,                      /* m_size */
  odbparser_methods,       /* methods */
  NULL,                    /* m_reload */
  NULL,                    /* m_traverse */
  NULL,                    /* m_clear */
  NULL                     /* m_free */
};


static PyObject *ErrorObject;

PyMODINIT_FUNC PyInit_odbparser(void) {
  PyObject *m, *d;

  /* Create the module and add the functions */
  m = PyModule_Create(&moduledef);

  /* Add some symbolic constants to the module */
  d = PyModule_GetDict(m);
  ErrorObject = PyUnicode_FromString("odbparser.error");
  PyDict_SetItemString(d, "error", ErrorObject);

  import_array();

  /* Check for errors */
  if (PyErr_Occurred())
    Py_FatalError("can't initialize module odbparser");

  return m;
}


/*
  Local Variables:
  mode: font-lock
  End:
*/
